### ISSUE TITLE
Consider renaming (`Re`)`EarlyBound` to `ReParam` and (`Re`)`LateBound` to `Bound`.

### ISSUE THREAD

**Issue by eddyb** [rust-lang/rust#70805]
This convention would be slightly less confusing due to aligning better with `Ty` and `ty::Const`.

On the other hand, the early-bound vs late-bound terminology is so baked-in that this might be hard.

cc @rust-lang/wg-traits @matthewjasper

**Comment by nikomatsakis** [comment]
Yes, I agree this is a terminology question to reconcile -- but more generally I think we want to ask ourselves is whether to revisit how we handle parameters altogether. In Chalk, there is no early/late bound distinction: there are only "bound things", and we use them for everything. We also always substitute bound things with placeholders or other free things when using them.

This is in contrast to the rustc approach of using `Param` and knowing from context whether it's to be treated as something *bound* or as a *placeholder*.

I think we should settle on one. I think the chalk approach is cleaner and would probably help prevent that weird behavior we get from time to time when someone forgets to apply a substitution (if we play our cards right, we could probably prevent that behavior *statically* by having two distinct `Interner` generic types, one for unsubstituted types and one for substituted types, although actually we can do that in either case I suppose).

On the other hand, the rustc approach means less overall substitution and types created, and this may have some memory/performance benefits.
